src/main/java/com/indeed/jiraactions/api/ApiCaller.java

```diff
@@ -3,178 +3,120 @@
-                final InputStream error = urlConnection.getErrorStream();
-                if(error != null) {
-                    final BufferedReader errorBr = new BufferedReader(new InputStreamReader(error));
-                    sb.append("\"Error Body\": \"");
-                    String line = errorBr.readLine();
-                    while (line != null) {
-                        sb.append(line).append(System.lineSeparator());
-                        line = errorBr.readLine();
-                    }
-                    errorBr.close();
-                } else {
-                    sb.append("Unable to open error stream for reading");
-                }
-                sb.append("\"");
+                sb.append("\"Code\": ").append(response.code()).append(",");
+                sb.append("\"Message\": \"").append(response.message()).append("\",");
+                sb.append("\"Error Body\": \"").append(responseBody.string()).append("\"");
+                sb.append('}');
+                sb.append('}');
+                throw new IOException(String.valueOf(response));
             }
-
-            sb.append("}");
-            sb.append("}");
-            log.error("Encountered connection error: " + sb);
+            return objectMapper.readTree(responseBody.string());
+        } catch (IOException e) {
             throw e;
-        } finally {
-            if (br != null) {
-                try {
-                    br.close();
-                } catch (final IOException ignored) {
-                }
-            }
         }
     }
 
-    private HttpsURLConnection getURLConnection(final String urlString) throws IOException {
-        final URL url = new URL(urlString);
-        final HttpsURLConnection urlConnection = (HttpsURLConnection) url.openConnection();
-        urlConnection.setRequestProperty("Authorization", authentication);
-
-        if(cookies.length() > 0) {
-            urlConnection.setRequestProperty("Cookie", cookies);
-        }
-        return urlConnection;
-    }
-
-    private void setCookies() {
-        final StringBuilder sb = new StringBuilder();
-        if(jsessionId != null) {
-            sb.append("JSESSIONID=").append(jsessionId);
-        }
-        if(upstream != null) {
-            if(sb.length() > 0) {
-                sb.append(";");
-            }
-            sb.append("upstream=").append(upstream);
-        }
-        cookies = sb.toString();
-    }
-
     private String getBasicAuth() {
         final String userPass = config.getJiraUsername() + ":" + config.getJiraPassword();
         final String basicAuth = "Basic " + new String(new Base64().encode(userPass.getBytes()));
         return basicAuth;
     }
+
+    private static CookieJar setupCookieJar() {
+        return new CookieJar() {
+            private final HashMap<String, List<Cookie>> cookieStore = new HashMap<>();
+
+            @Override
+            public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
```

Comment:
Does this work? It looks like `HttpUrl` contains path segments, which means every URL will be unique. One hack would be to simply use a single List of cookies, since we know we're only connecting to a single domain and all the cookies are safe to share.

Also consider something like this: https://github.com/franmontiel/PersistentCookieJar or other alternatives in this thread: https://github.com/square/okhttp/issues/2890

Reply:
Tried importing the PersistentCookieJar, but it was in a different repository and was acting up. I'm gonna try going with the single List of cookies approach.

Reply:
Looks like PersistentCookieJar is too tightly coupled on Android stuff for us anyway.

Depending on how OkHttp uses the CookieJar, we may want a Set<> instead of a List<>.